### DISCUSSION TITLE
New attribute/directive for Bundle-ActivationPolicy that controls SCR behaviour?

### DISCUSSION THREAD

**Discussion by laeubi** [osgi/osgi#590] [Ideas]
Currently DS is activating components either when

1. the bundle is `ACTIVE`
2. the bundle is `STARTING` (having `Bundle-ActivationPolicy: lazy`)

in the past this has produced some problems in the context of Eclipse Platform, because here some functionality relies on the fact that the bundle is activates as late as possible (e.g. only when you activate a view the first time) and has using `Bundle-ActivationPolicy: lazy` for that, but if one has only one `immediate` component this enables the plugin always eagerly.

This currently is one of the biggest push-backs to using DS in platform instead of Activators (FYI @vogella) because currently this is only achievable with activators.

I therefore like to discuss if it would be feasible to have a new attribute/directive for the `Bundle-ActivationPolicy` (but I would be open for other solutions as well e.g. having a dedicated header or something in the component.xml) to better control the behavior, e.g I can think about a `ds-policy` in the following way:

- `Bundle-ActivationPolicy: lazy;ds-policy:=eager` (the default) behaves as of today, any lazy component will be activated if requested and if there is an immediate component
- `Bundle-ActivationPolicy: lazy;ds-policy:=ignore` behaves as if the bundle is *not* lazy, that means it will only activate any component if the bundle is finally `ACTIVE`
 - `Bundle-ActivationPolicy: lazy;ds-policy:=lazy` only activates lazy components if requested and immediate components once the bundles becomes finally `ACTIVE`.

**Comment by juergen-albert** [comment]
I stumbled upon this issue a while back as well. After talking to @tjwatson I found out, that this is not an osgi but an equinox issue. There is somewhere a small specific peace of code that puts certain bundles in the starting stage, even though they are not in the process of being started. @tjwatson can tell you more about that. 

This causes the issues you see above. There is even an option to disable this behaviour if I remember correctly.

Without this strange behaviour , this is a non issue, as DS acts as desired.

Besides that, we now have the [Satisfying Condition](https://docs.osgi.org/specification/osgi.cmpn/8.0.0/service.component.html#service.component-satisfying.condition) to address such issues. A bundle can register a Condition service at the end of its Activation and all services that target that condition will not be activated until it becomes enabled. 
When I get around to finalizing the Condtition Factory specification, we hopefully have a part included, that registers such conditions for you dependent on bundles available/active/inactive, so this can be done via configuration and not via code.

Hopefully TJ can tell us more about the Eclipism you described here.

**Reply by laeubi** [reply]
[from the spec](https://docs.osgi.org/specification/osgi.cmpn/7.0.0/service.component.html#d0e36881)

> An immediate component is activated as soon as its dependencies are satisfied. If an immediate component has no dependencies

So assume I have an immediate component with no dependencies then loading of the service class, [then further on](https://docs.osgi.org/specification/osgi.cmpn/7.0.0/service.component.html#service.component-activation)

> Activating a component configuration consists of the following steps:
> * Load the component implementation class.

So for me it reads as if this will trigger then activation of the bundle immediately as a class is loaded from the bundle.

**Reply by timothyjward** [reply]
> So for me it reads as if this will trigger then activation of the bundle immediately as a class is loaded from the bundle.

This is what lazy activation means. 

> So assume I have an immediate component with no dependencies

As of the latest DS all components have at least one dependency on a condition. By default this is a condition that is always available, but it can be changed (either statically in the component definition or in configuration) to be one that is registered later. This is why @juergen-albert is saying that we don’t need anything added to the specification - you can use a custom condition to delay activation as long as you need.

**Reply by laeubi** [reply]
Having a custom condition does not really seems something one likes here as it is really complicated to implement, if that's required it would probably easier to not use DS at all as it requires to implement similar code that an activator already would offer. Beside that, it requires each component author to know about and take care to choose the "right" condition.

What I like to control is (on the bundle level) that DS do not activate immediate components on lazy bundles unless the bundle becomes active by some other mean.

**Reply by laeubi** [reply]
And just out of curiosity: Just in case I would supply such a condition (what has to be a service) I would either need an `Activator` (what a bit defeats the purpose) or the Condition itself might already trigger activation of my bundle... ;-)

**Reply by juergen-albert** [reply]
Which is what DS is not doing. The case you mentioned initially is only possible, because of the "strange" behaviour of equinox. 
Usually bundles are started quickly and DS reads and acts on the component description after the activation is done.

**Reply by laeubi** [reply]
> Which is what DS is not doing. 

Can you explain what DS "is not doing"? If a bundle is marked as `ActivationPolicy: lazy` it activates components, and as immediate ones are activated instantly this triggers the loading, so I don't see what Equinox has to do here.

> Usually bundles are started quickly and DS reads and acts on the component description after the activation is done.

The point is that the bundle should **not** be started but is implicitly because DS loads a class from it.

**Reply by juergen-albert** [reply]
> And just out of curiosity: Just in case I would supply such a condition (what has to be a service) I would either need an `Activator` (what a bit defeats the purpose) or the Condition itself might already trigger activation of my bundle... ;-)

Here I'd like to point to the planed condition factory, which can take care of such things. I hope to get something ready to discuss here soon.

**Reply by laeubi** [reply]
But there must be some kind of code implementing it, listen for the condition to become true and register something, al of this requires an active bundle (what I want to avoid here...).

**Reply by juergen-albert** [reply]
> Can you explain what DS "is not doing"? If a bundle is marked as ActivationPolicy: lazy it activates components, and as immediate ones are activated instantly this triggers the loading, so I don't see what Equinox has to do here.

A bundle marked as lazy is loaded and will remain in the state of resolved (if it does resolve) until someone manually starts it. That you or vogella have seen a different behaviour is caused by the behaviour of equinox to set lazy bundles under certain condition's (TJ again for details) in the state of starting even if they are actually not started at that time. This causes DS to start processing them.

**Reply by laeubi** [reply]
Whole Eclipse IDE actually depends on the behavior then, there is even a setting in PDE that enable lazy-activation if DS Components are used (because usually users don't have a way to "start" things otherwise), changing this will likely break a lot of things inside Eclipse.

So maybe its just a matter of terminology here, to summarize then what I want is:

1. Bundle is in `STARTING` state (no class was loaded from it)
2. I want DS to keep either all components alone unless it is `ACTIVE` (case 2) or at least all immediate one (case 3)

For me that's not really something I would express by a _condition_ (and I have no idea how one could construct such condition without major efforts) it is a change in _behavior_ of the DS, and for DS there is a clear observable state like shown here:

https://docs.osgi.org/specification/osgi.core/8.0.0/framework.lifecycle.html#i3270475

So if DS do not read `Bundle-ActivationPolicy` then maybe a separate header to control this would be more appropriate, but it feels natural to use that header as it obviously is part of the behavioral change, and looking at the state diagram DS actually **must** listen to `STARTING` event already because otherwise it wont notice the bundle/components at all.

**Reply by timothyjward** [reply]
> al of this requires an active bundle (what I want to avoid here...).

I thought you wanted to avoid accessing eclipse components before they initialise? You can load as many classes as you like from your own bundle, and other bundles, just don’t access the components .

**Reply by laeubi** [reply]
> I thought you wanted to avoid accessing eclipse components before they initialise? 

Formally I want to avoid that DS loads classes from a bundles unless it is in state `ACTIVE` (case 2)

> You can load as many classes as you like from your own bundle, and other bundles, just don’t access the components .

This might happen and then the bundle becomes `ACTIVE` (when marked with lazy and started).

Just assume this very common eclipse (IDE/RCP) case:

1. I write a plugin that offers a new cool view, well at least I think so but other users might find it not that important and *don't show it by default*
2. To properly function, I need to access other services and if these are using the whiteboard-pattern I need to register services, so all the usual way we know
3. Usually such a bundle is marked with `Bundle-ActivationPolicy: lazy` and resulting in be in `STARTING` state (@juergen-albert mentioned it might be a sideffect but that's how it is as of today and I can hardly think that this could be changed)
4. Also quite common for such an eclipse-plugin (aka bundle), that it contains some metadatafiles, this could be `plugin.xml` for extension points, `fragment.xml` for E4, or even helper classes called by other bundles (e.g. `JavaCore.<....>`)
5. Last but not least, such plugin often contains `BundleActivators` (or even worse `PluginActivator`) that perform the acquire and register of services (what I'd like to replace by DS and recommend for new design).
6. Now by default everything "works" nicely, because by bundle is "inactive" (no services consumed, no services registered, no other bundles triggered for activation)
7. If the users ask Eclipse to actually show the view, the view class gets loaded, what triggering the Activator to run, register services, gather services and other classes that gain triggers other bundles/services to load/start and so on and my bundle becomes `ACTIVE`.

Now assume I add DS to the mix, and for example either add an immediate component, or a "whiteboard component" (e.g. `EventListener`), the whole thing "breaks" (in Eclipse IDE/RCP terms!) because my bundle is now becoming `ACTIVE` immediately with all the side effects (other bundles activated, services started and so on), **without the view actually shown somewhere**.

This can have quite dramatic consequences as it can lead to a massive "activation rush" with UI hanging / unresponsive, triggering some insane start-order dependency of bad written components and so on (a usual Eclipse can easily contain thousands of bundles!), leading to the problem that a good idea tor improve things (getting away from clumsy Activators with static access all over the code) suddenly "breaks" things that "worked for years" and people like to stay at status-quo.

Of course many issues *can* be avoided other ways (and I already get some progress here over the past releases), but this is heavily hold back as it require very careful tweakings and non trivial refactorings, while actually all one needs is that DS leaves the bundle alone in STARTING state to allow for a smooth transition :-)

The [spec actually mentions this case](https://docs.osgi.org/specification/osgi.core/8.0.0/framework.lifecycle.html#i3270394) as well here:

> For larger systems, this greedy strategy can significantly increase startup times and unnecessarily increase the memory footprint. Mechanisms such as the Service Component Runtime and activation policies can mitigate these problems. 

But the (currently not controllable) behavior of DS currently defeats the whole lazy activation and therefore  "significantly increase startup times and unnecessarily increase the memory footprint".

**Reply by juergen-albert** [reply]
Per Specification, a Bundle should only shortly enter the state of STARTING for the time its Activator is running. The Spec states somewhere that long running operations are to be avoided in that phase. Equinox has this specific non spec behaviour that it puts some bundles (not exactly sure what the trigger is) in the state of starting, without actually starting it (if the bundle policy is lazy). 

Thus the behaviour your request seem to come from, is from a "misbehaviour" of equinox. It always confused the hell out of me as well at the time. The main issue remains, that equinox has an option witch is active by default, that deviates from the specified requirement. Thus I don't believe that it warents a change in the Spec. It should be rather discussed if this can be changed in equinox instead, as this was something p2 relied on way back when, if I remember correctly. 

In Felix e.g. you will never see a Bundle in the state of starting for longer as a few milliseconds, as they are either resolved, installed or stoped.

**Reply by tjwatson** [reply]
> Per Specification, a Bundle should only shortly enter the state of STARTING for the time its Activator is running. The Spec states somewhere that long running operations are to be avoided in that phase. Equinox has this specific non spec behaviour that it puts some bundles (not exactly sure what the trigger is) in the state of starting, without actually starting it (if the bundle policy is lazy).

I think this has nothing to do with the discussion at hand.  By default Equinox will put lazy activation policy bundles into the STARTING state as if they were explicitly started with the activation policy.  But from that point they behave just like a lazy activated bundle on Felix that was explicitly started with the activation policy.

**Comment by laeubi** [comment]
> In Felix e.g. you will never see a Bundle in the state of starting for longer as a few milliseconds, as they are either resolved, installed or stoped.

I think this violates the spec then [see this graph](https://docs.osgi.org/specification/osgi.core/8.0.0/framework.lifecycle.html#i3270475):

![](https://docs.osgi.org/specification/osgi.core/8.0.0/images/004-eager-lazy-activation.png)

it says a bundle has to stay in `STARTING` until a class is loaded, [also described here](https://docs.osgi.org/specification/osgi.core/8.0.0/framework.lifecycle.html#i3270445):

> When a bundle is started using a lazy activation policy, the following steps must be taken:
> 
> - A Bundle Context is created for the bundle.
> - The bundle state is moved to the [STARTING](https://docs.osgi.org/specification/osgi.core/8.0.0/framework.api.html#org.osgi.framework.Bundle.STARTING) state.
> - The [LAZY_ACTIVATION](https://docs.osgi.org/specification/osgi.core/8.0.0/framework.api.html#org.osgi.framework.BundleEvent.LAZY_ACTIVATION) event is fired.
> - **The system waits for a class load from the bundle to occur.**
> - The normal [STARTING](https://docs.osgi.org/specification/osgi.core/8.0.0/framework.api.html#org.osgi.framework.BundleEvent.STARTING) event is fired.
> - The bundle is activated.
> - The bundle state is moved to [ACTIVE](https://docs.osgi.org/specification/osgi.core/8.0.0/framework.api.html#org.osgi.framework.Bundle.ACTIVE).
> - The [STARTED](https://docs.osgi.org/specification/osgi.core/8.0.0/framework.api.html#org.osgi.framework.BundleEvent.STARTED) event is fired.

**Comment by timothyjward** [comment]
As discussed on today's call, it's not appropriate to use the following:

    Bundle-ActivationPolicy: lazy;ds-policy:=eager

The `Bundle-ActivationPolicy` header is owned by the core specification and it should not have to understand anything about DS. If we need to put an attribute/directive anywhere it should be on `Service-Component` or a new header which is specific to the DS specification.

**Comment by tjwatson** [comment]
Why have an immediate component if you do not want it eagerly activated?  If a bundle really wants to have lazy behavior I would think it would want to have no immediate components.

**Reply by laeubi** [reply]
This also happens for "non eager" component, this can also happen for e.g. whiteboard services where one only want the bundle to be activated by some other means (e.g. extension point), for example you can look here to se how this causes issues:

- https://github.com/eclipse-pde/eclipse.pde/pull/542
- https://github.com/eclipse-pde/eclipse.pde/pull/536#issuecomment-1489733174

so an immediate component is just a replacement for a service currently registered in an `Activator` (for what it works fine) ...

**Comment by timothyjward** [comment]
This will be gathered as a requirement for DS 1.6 in the requirements/590 branch.

**Reply by laeubi** [reply]
> This will be gathered as a requirement for DS 1.6 in the requirements/590 branch.

@timothyjward I just wanted to make sure, is this already worked on or is there any further action needed (possibly by me?).

**Reply by timothyjward** [reply]
@laeubi - you have the pen on this one I think!

**Reply by laeubi** [reply]
@timothyjward I have now created
- https://github.com/osgi/osgi/pull/633
